sound/soc/sof/ipc.c

```diff
@@ -279,6 +280,57 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
 	/* now wait for completion */
 	return tx_wait_done(ipc, msg, reply_data);
 }
+
+/* send IPC message from host to DSP */
+int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
+			void *msg_data, size_t msg_bytes, void *reply_data,
+			size_t reply_bytes)
+{
+	struct sof_ipc_large_hdr *hdr;
+	uint32_t num_msg = 0;
+	uint32_t offset = 0;
+	uint32_t send_bytes = 0;
+	int ret = -1;
+	int size;
+	int i;
+
+	if (msg_bytes <= SOF_IPC_MSG_MAX_SIZE)
+		return sof_ipc_tx_msg(ipc, header, msg_data, msg_bytes,
+				      reply_data, reply_bytes);
+
+	/* if bigger than max size, chop to smaller pieces */
+	dev_dbg(ipc->sdev->dev, "sending large ipc size %zu\n", msg_bytes);
+	size = SOF_IPC_MSG_MAX_SIZE - sizeof(struct sof_ipc_large_hdr);
+	num_msg = (msg_bytes + size - 1) / size;
+
+	hdr = (struct sof_ipc_large_hdr *)
+		kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!hdr)
+		return -ENOMEM;
+
+	hdr->hdr.cmd = SOF_IPC_GLB_LARGE;
+	hdr->cmd = header;
+	hdr->count = num_msg;
+
+	for (i = 0; i < num_msg; i++) {
+		hdr->id = i;
+		send_bytes = msg_bytes > size ? size : msg_bytes;
+		memcpy(hdr->data, msg_data + offset, send_bytes);
+		hdr->hdr.size = send_bytes;
+		ret = sof_ipc_tx_msg(ipc, SOF_IPC_GLB_LARGE, hdr, send_bytes +
+				     sizeof(struct sof_ipc_large_hdr),
+				     reply_data, reply_bytes);
+		if (ret < 0)
```

Comment:
why not use multiple regular IPC transfers, I don't get the need for a new type of transfers.

Reply:
so actually I renamed the old tx_message to tx_msg, so this is sending multiple "old style" max size ipc transfers (they are just framed inside new struct to be able to gather up on dsp side). Smaller than max transfer size messages are going just as before.  This way I didn't have to touch topology.c.